### DISCUSSION TITLE
下面这个子类的实例化过程是什么顺序

### DISCUSSION THREAD

**Discussion by Lining0812** [BYJRK/DotNet-Discussions#353] [C# 技术交流]
基类构造器中调用了一个需要子类实现的纯虚方法，这个子类的实例化过程是什么样的
    using System;

    namespace ConsoleApp2
    {
        class Program
        {
            static void Main(string[] args)
            {
                PrimaryStu lihua = new PrimaryStu("Li");
            }
        }

        public abstract class Student
        {
            protected Student(string name)
            {
                Name = name;
                Creat();
            }

            public string Name { get; set; }
            public virtual void Creat()
            {
                Console.WriteLine("实例化。。。");
                Regist();
                Console.WriteLine("注册了吗。。。");
            }
            public abstract void Regist();
        }

        public class PrimaryStu : Student
        {
            public PrimaryStu(string name):base(name)
            {
            }
            public override void Regist()
            {
                Console.WriteLine("注册信息。。。");
            }
        }
    }

**Comment by CodingOctocat** [comment]
Console.WriteLine() 里面多放一个 this.GetType() 就知道了

**Reply by Lining0812** [reply]
有道理，居然都是子类神奇。子类实例化时会调用父类的构造器原来是这个意思

**Comment by ax020913** [comment]
关于virtual，override继承的内容了

**Comment by hjslovehr** [comment]
1. 先走子类的构造函数声明那一行，那一行不会进到函数体，这一步只是申请内存
2. 然后发现有基类，跳到基类构造函数那一行，如果基类还有父类，继续找往上找。没有就进入基类构造函数，执行成员变量初始化
3. 然后调用 Creat() 
4. Creat() 内部调用抽象方法 Regist()，此时寻找对应子类的 Regist() 方法。
5. Regist() 方法退出
6. Creat() 方法退出
7. 基类构造函数结束
8. 回到子类构造函数继续执行
